rllib/agents/impala/vtrace_torch_policy.py

```diff
@@ -251,51 +163,198 @@ def make_time_major(policy, seq_lens, tensor, drop_last=False):
     return res
 
 
-def stats(policy: Policy, train_batch: SampleBatch) -> Dict[str, Any]:
+class VTraceOptimizer:
+    """Optimizer function for VTrace torch policies."""
 
-    return {
-        "cur_lr": policy.cur_lr,
-        "total_loss": torch.mean(torch.stack(policy.get_tower_stats("total_loss"))),
-        "policy_loss": torch.mean(torch.stack(policy.get_tower_stats("pi_loss"))),
-        "entropy": torch.mean(torch.stack(policy.get_tower_stats("mean_entropy"))),
-        "entropy_coeff": policy.entropy_coeff,
-        "var_gnorm": global_norm(policy.model.trainable_variables()),
-        "vf_loss": torch.mean(torch.stack(policy.get_tower_stats("vf_loss"))),
-        "vf_explained_var": torch.mean(
-            torch.stack(policy.get_tower_stats("vf_explained_var"))
-        ),
-    }
+    def __init__(self):
+        pass
 
+    @override(TorchPolicyV2)
+    def optimizer(
+        self,
+    ) -> Union[List["torch.optim.Optimizer"], "torch.optim.Optimizer"]:
+        if self.config["opt_type"] == "adam":
+            return torch.optim.Adam(params=self.model.parameters(), lr=self.cur_lr)
+        else:
+            return torch.optim.RMSprop(
+                params=self.model.parameters(),
+                lr=self.cur_lr,
+                weight_decay=self.config["decay"],
+                momentum=self.config["momentum"],
+                eps=self.config["epsilon"],
+            )
+
+
+# VTrace mixins are placed in front of more general mixins to make sure
+# their functions like optimizer() overrides all the other implementations
+# (e.g., LearningRateSchedule.optimizer())
+class VTraceTorchPolicy(
+    GradClippingMixin,
```

Comment:
Note (Doesn't have to be in this PR): We should replace mixins - if possible - with method overrides. Most mixins don't have state but only define one method. It would be clearer for the users mental model to simply override the method in the Policy sub-class and then use an imported utility function in there. Mixins are scary, you never know, which one to initialize first to not step over some stuff in the other ones.

Reply:
done.
I left compute_gradients() in tf_mixins.py, which tbh may not be the best place.
added a TODO about it.

Reply:
otherwise, GAE and ComputeAndClipGrad are not MixIns anynore.